### PR TITLE
Feature/sign in

### PR DESCRIPTION
**Pull request by oyakovytskyi**
## Describe your changes

- Email Validation: Implemented email validation using react-hook-forms to improve form accuracy and user experience.
- Component Refactoring: Added TextInputRef in the component, allowing for correct form handling with react-hook-forms.
- Router Configuration: Updated router settings to align with the latest project changes.
- State Persistence: Integrated redux-persist in Redux, enabling persistent state management across sessions.
- API Requests for Login: Created userActions for backend API login requests, ensuring smoother login flow.
- 404 Page Redirection: Added a notFound page to redirect logged-in users to the homepage, enhancing navigation.
- TypeScript Interfaces: Established a dedicated folder for interfaces, aiding in organized code structure and better type - management.
- Minor Adjustments: Made small changes to imports, theme styling, and text for better code consistency.

## Issue ticket code (and/or) and link

- [Link to JIRA ticket](https://codecraftersintership.atlassian.net/jira/software/projects/SCRUM/boards/1?selectedIssue=SCRUM-4)

### **General**

- [x] Assigned myself to the PR
- [x] Assigned the appropriate labels to the PR
- [x] Assigned the appropriate reviewers to the PR
- [ ] Updated the documentation
- [x] Performed a self-review of my code
- [x] Types for input and output parameters
- [x] Don't have "any" on my code
- [x] Used the try/catch pattern for error handling
- [x] Don't have magic numbers
- [x] Compare only with constants not with strings
- [x] No ternary operator inside the ternary operator
- [x] Don't have commented code
- [x] no links in the code, env links should be in env file (for example: server url), constant links (for example default avatar URL) should be in constant file.
- [x] Used camelCase for variables and functions
- [x] Date and time formats are on the constants
- [x] Functions are public only if it's used outside the class
- [x] No hardcoded values
- [ ] covered by tests
- [x] Check your commit messages meet the [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/).

### Frontend

- [x] Components and business logic are separated
- [x] Colors, Font Size, and Font Name is on the theme or in the constants
- [x] No text in the components, use i18n approach
- [x] No inline styles
- [x] Imports are absolute
- [x] Attach a screenshot if PR has visual changes.

photo_2024-11-12_16-11-28
photo_2024-11-12_20-28-52
photo_2024-11-12_20-29-05
After 5 attempts user unable to press button and see error.
photo_2024-11-12_20-29-10
Pop-up message after validation email.

### CODE REVIEW

File: src/api/userActions.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+import axiosInstance from '@/utils/axiosInstance';
+import { setAccessToken } from '@/store/slices/authSlice';
+
+export const sendLoginLink = (email: string) => async (): Promise<boolean> => {
+  try {
+    const response = await axiosInstance.get(`/auth/${email}`);
+    if (response.data.status) {
+      toast.success(
+        `Message successfully sent to ${email}, please check your email.`
```
Comment (lyudad) [inline src/api/userActions.ts]: move to json file

File: src/pages/SignIn/useSignIn.ts [hunk 2]
```diff
@@ -1,77 +1,92 @@
...
+  dispatch: AppDispatch
+): UseSignInReturnType => {
   const { t } = useTranslation();
-  const location = useLocation();
-  const navigate = useNavigate();
-  const dispatch = useDispatch();
+  const [attemptCount, setAttemptCount] = useState(0);
+  const [isLocked, setIsLocked] = useState(false);
+
+  const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$/;
```
Comment (lyudad) [inline src/pages/SignIn/useSignIn.ts]: move above the hook

File: src/pages/SignIn/useSignIn.ts [hunk 3]
```diff
@@ -1,77 +1,92 @@
...
-  handleEmailChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
-  handleSubmit: () => void;
-}
-
-const useSignIn = (): UseSignInReturnType => {
+const useSignIn = (
+  navigate: ReturnType<typeof useNavigate>,
+  dispatch: AppDispatch
+): UseSignInReturnType => {
   const { t } = useTranslation();
```
Comment (lyudad) [inline src/pages/SignIn/useSignIn.ts]: under use state

File: src/pages/components/TextInputRef/index.tsx [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import { forwardRef } from 'react';
+
+import TextInput, { TextInputProps } from '@/components/TextInput';
+
+const TextInputWithRef = forwardRef<HTMLInputElement, TextInputProps>(
```
Comment (lyudad) [inline src/pages/components/TextInputRef/index.tsx:5]: add a comment above the component //add ref to prevent ref error

**Review by lyudad — APPROVED** [review]

File: src/api/userActions.ts (hunk 1 above)
Comment (oyakovytskyi) [reply src/api/userActions.ts]: fixed

File: src/pages/SignIn/useSignIn.ts (hunk 2 above)
Comment (oyakovytskyi) [reply src/pages/SignIn/useSignIn.ts]: done

File: src/pages/SignIn/useSignIn.ts (hunk 3 above)
Comment (oyakovytskyi) [reply src/pages/SignIn/useSignIn.ts]: ok, done

File: src/pages/components/TextInputRef/index.tsx (hunk 4 above)
Comment (oyakovytskyi) [reply src/pages/components/TextInputRef/index.tsx:5]: comment added

**Pull request merged**